### ISSUE TITLE
DirectoryInfo.GetFiles (System.String searchPattern, System.IO.SearchOption searchOption) crashes

### ISSUE THREAD

**Issue by chamons** [mono/mono#17171]
_From @ivanicin on Wed, 02 Oct 2019 20:08:05 GMT_

### Steps to Reproduce

I use this line of code: 
`var files = DirectoryInfo.GetFiles ("*.*", System.IO.SearchOption.AllDirectories);`

For one user (who seems to have a lot of files created inside), this crashes with SIGABRT: Operation not permitted, and from the stacktrace you can see that the bug is coming from Xamarin code, not from my code, here it is:
 
    FileSystemEnumerator`1[TResult].CreateDirectoryHandle (System.String path, System.Boolean ignoreNotFound)
    FileSystemEnumerator`1[TResult].DequeueNextDirectory ()
    FileSystemEnumerator`1[TResult].DirectoryFinished ()
    FileSystemEnumerator`1[TResult].FindNextEntry (System.Byte* entryBufferPtr, System.Int32 bufferLength)
    FileSystemEnumerator`1[TResult].MoveNext ()
    LargeArrayBuilder`1[T].AddRange (System.Collections.Generic.IEnumerable`1[T] items)
    EnumerableHelpers.ToArray[T] (System.Collections.Generic.IEnumerable`1[T] source)
    MonoLinqHelper.ToArray[T] (System.Collections.Generic.IEnumerable`1[T] source)
    DirectoryInfo.GetFiles (System.String searchPattern, System.IO.EnumerationOptions enumerationOptions)
    DirectoryInfo.GetFiles (System.String searchPattern, System.IO.SearchOption searchOption)
    SettingsController+<>c__DisplayClass2_0.<ViewDidLoad>b__5 ()
    UIGestureRecognizer+ParameterlessDispatch.Activated ()
    (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
    UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate)
    UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName)
    Application.Main (System.String[] args)

### Expected Behavior
It should work, if there is some problem it should raise an exception, not to crash inside the method
### Actual Behavior
It crashes inside the method while trying to enumerate subdirectories
### Environment

```
Microsoft Visual Studio Community 2019
Version 16.2.5
VisualStudio.16.Release/16.2.5+29306.81
Microsoft .NET Framework
Version 4.8.03752

Installed Version: Community

Visual C++ 2019   00435-60000-00000-AA588
Microsoft Visual C++ 2019

Application Insights Tools for Visual Studio Package   9.1.00611.1
Application Insights Tools for Visual Studio

ASP.NET and Web Tools 2019   16.2.292.25104
ASP.NET and Web Tools 2019

ASP.NET Web Frameworks and Tools 2019   16.2.292.25104
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   16.2.292.25104
Azure App Service Tools v3.0.0

Azure Functions and Web Jobs Tools   16.2.292.25104
Azure Functions and Web Jobs Tools

C# Tools   3.2.1-beta4-19408-03+2fc6a04980f800c59e8ede97e6ae294ff47d666e
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Extensibility Message Bus   1.2.0 (master@8b56e20)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

IntelliCode Extension   1.0
IntelliCode Visual Studio Extension Detailed Info

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 0x10 - v2.9.20626.2

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   1.0
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

Mono Debugging for Visual Studio   16.2.6 (4cfc7c3)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   5.2.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

SQL Server Data Tools   16.0.61906.28070
Microsoft SQL Server Data Tools

TypeScript Tools   16.0.10627.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   3.2.1-beta4-19408-03+2fc6a04980f800c59e8ede97e6ae294ff47d666e
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools 10.4 for F# 4.6   16.2.0-beta.19321.1+a24d94ecf97d0d69d4fbe6b8b10cd1f97737fff4
Microsoft Visual F# Tools 10.4 for F# 4.6

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

Xamarin   16.2.0.149 (d16-2-xcode11@c5da8fd8b)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   16.2.0.375 (remotes/origin/d16-2@357d38ef4)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   16.3.118 (c29ca2a)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   9.4.1.1 (d16-2/cec9eb4)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: mono/mono/2019-02@e6f5369c2d2
    Java.Interop: xamarin/java.interop/d16-2@d64ada5
    LibZipSharp: grendello/LibZipSharp/d16-2@caa0c74
    LibZip: nih-at/libzip/rel-1-5-1@b95cf3f
    ProGuard: xamarin/proguard/master@905836d
    SQLite: xamarin/sqlite/3.27.1@8212a2d
    Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-2@6f6c969


Xamarin.iOS and Xamarin.Mac SDK   12.99.2.40 (d883feb)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.
```

### Example Project (If Possible)

_Copied from original issue xamarin/xamarin-macios#7159_

**Comment by chamons** [comment]
_From @ivanicin on Thu, 03 Oct 2019 15:18:22 GMT_

According to this: https://github.com/dotnet/corefx/issues/39768#issuecomment-515169831 , this could be avoided if the method used Directory.GetDirectories instead of Directory.EnumerateDirectories. As this method is clearly labeled as 'Get' and not 'Enumerate', this needs to be changed accordingly.

**Comment by baulig** [comment]
I suppose we can close this here as well?

**Comment by baulig** [comment]
It looked to me like the issue was fixed because it linked to another closed issue.  Maybe I was wrong about that?

**Comment by chamons** [comment]
To clarify - This turns out to not be a bug but mis-use of the API in question.

Here's the text from the [linked corefx bug comment](https://github.com/dotnet/corefx/issues/39768#issuecomment-515169831) which can explain it better than I can:

> EnumerateDirectories needs to open a handle and keep it open while enumerating; the alternative would be to open the handle, cache everything into a data structure, close the handle, and then return the data structure. But that's exactly what GetDirectories does, and EnumerateDirectories was introduced later specifically to avoid these overheads.
> 
> If a developer actually wants the caching behavior, they can use GetDirectories (or use ToArray). If a developer wants the recursive behavior, they can specify so with SearchOption and let the implementation handle everything. If the developer is explicitly calling other things while holding open the handle, well, that's what they get.

**Comment by ivanicin** [comment]
@chamons I guess that you don't understand my explanation - or I don't understand your explanation, but let's clear it out if it is the first thing.

So here is the story line of the problem:

- DirectoryInfo.GetFiles (System.String searchPattern, System.IO.SearchOption searchOption) crashes 
- according to the stack trace, this happens because its implementation uses Directory.EnumerateDirectories which causes the crash
- it is known that Directory.EnumerateDirectories causes such a problem when being misused
- a simple logic leads to conclusion that DirectoryInfo.GetFiles (System.String searchPattern, System.IO.SearchOption searchOption) misuses Directory.EnumrateDirectories in its implementation and that improper implementation causes the crash. It needs to be fixed in .NET Core.

Looking above I guess that I haven't been clear in explaining as this seems very obvious when understood properly. Still I might be wrong.

**Comment by chamons** [comment]
I'm just forwarding a response from the runtime team to make it more clear. I'm not on the mono team (directly).

@baulig - Thoughts on ^? Does it change the bug status?

**Comment by steveisok** [comment]
@ivanicin What does your directory structure look like?  And when you say a lot of files, how many?

I'm going to reopen for the time being because I want to make sure we understand your scenario better.

**Comment by ivanicin** [comment]
@steveisok it is somewhat ironic that I've got to this problem by trying to answer that question. 

So the story goes that one user reported that he has 3 GB iCloud backup of data in my app. Truth to be told I still don't have the exact idea how one can do it, and I need to get all files and their size to properly answer the question. 

I am pretty sure that he has at least 1000 files, but whether there is much more who can tell without getting the data.

Going back to the problem while I understand why you would like this information, I also don't understand why it would be necessary from your side.

If you don't have any complaints on the bullets I've pointed above, it is a bug that needs to be fixed. If it is correct then it is very easy to be fixed, just changing the method `Directory.EnumerateDirectories ` to `Directory.GetDirectories ` inside the implementation of `DirectoryInfo.GetFiles(System.String searchPattern, System.IO.SearchOption searchOption)` will fix the issue.

**Comment by baulig** [comment]
Just tested this and couldn't reproduce it with my local Mono checkout (about 14k files).  But let's investigate.

**Comment by baulig** [comment]
Actually, it's 142k files not 14k.  Tested with several different versions with Mono after setting `ulimit -n 5`.

**Comment by ivanicin** [comment]
@baulig there is an extensive discussion on linked comment and some more bugs on how `Directory.EnumerateDirectories` causes the crash, so you should first insure that that method would crash. I am not sure what is exactly required, but as it is extensively discussed there it should be quite reproducible in certain cases.

You probably don't have enough subdirectories, that would be my guess.

**Comment by ivanicin** [comment]
@baulig what platform do you test on? Personally I report this for iOS, but it seems from the linked bug that macOS is affected too, though I can't personally confirm that. The threshold may be different though.

**Comment by baulig** [comment]
I'm testing on macOS.  But either way, we're using the CoreFX implementation, so we'd have to upstream any potential fix.

However, it looks like they made a few changes in their `release/3.0` branch that are not in our fork - although I couldn't find anything in particular in the commit messages that would hint at this specific problem.

**Comment by ivanicin** [comment]
Also note that I am in the contact with the user who has the problem so I can run any kind of code that would bring some more info. Each cycle will take at least a week, but it is possible.

**Comment by steveisok** [comment]
@ivanicin I'm trying to make sure we see it the same way.  

@baulig You could try using the angular template in Visual Studio.  That'll likely create more than enough in the node_modules directory.

**Comment by steveisok** [comment]
I did some digging and there are a couple points to cement:

1.) `GetFiles("*.*", SearchOption.AllDirectories)` is calling the same function as `EnumerateFiles` except that it's dumping the results to a FileInfo array.  That is supposed to guarantee all underlying handles are closed.  

2.) The routine to recurse through all the directories and grab all files has only 1 directory handle open at a time.  All child directories in the directory being processed are queued.  When at the end of the current directory, the current directory handle is closed, the next path is popped off the queue, and the process repeats. 

https://github.com/mono/corefx/blob/10a41e9f9828b3e299215406eadfde2ee4c9b29f/src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.Unix.cs#L120-L170

Which ends up calling at the end of a given directory:

https://github.com/mono/corefx/blob/10a41e9f9828b3e299215406eadfde2ee4c9b29f/src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.cs#L45-L62

I wanted to test this out on desktop mono first to see if there was some common problem across our profiles.  For my use case, I used `dotnet new angular` and the resulting node_modules folder after you build it (50K+ files). Unfortunately, it worked as I expected.

@chamons I wonder if some of the handles in iOS aren't being truly closed?  That's the only other reasonable explanation I can come up with.

**Comment by ivanicin** [comment]
At the very start someone referenced my issue from the third issue and pointed out that those issues are likely linked which sounded reasonable to me, I guess it does sound reasonable as you have tried it too.

However, it could be that the problem is in something else, like that there is some directory that can't be accessed, though I am not aware that such thing is possible in the app's directory on iOS. I guess that maybe there is some alternative explanation why this happens.

**Comment by ivanicin** [comment]
Some further info - the user has around 12k files in the library, I've got them by getting the files from two directories (without recursive search) that I have created and it worked. However the search above also was supposed to return cache and temp content, I would guess that the cache may contain a lot of files,  (but it is hard to be more precise as it is a webview cache).

So overall it is a weird issue and I agree it is the best to be closed, but if you think there is some way to get some data and actually fix the bug (if any), I am there to update the app on TestFlight and get the data from the user.
